mmdata.c

```diff
@@ -1137,8 +1118,7 @@ long nmbrRevInstr(long start_position, nmbrString *string1,
 
 
 /* Converts nmbrString to a vstring with one space between tokens */
-vstring nmbrCvtMToVString(nmbrString *s)
-{
+temp_vstring nmbrCvtMToVString(const nmbrString *s) {
   long i, j, outputLen, mstrLen;
   vstring tmpStr = "";
```

Comment:
Why do you avoid use of vstringdef here?
Suggest to move the declaration of ptr and ptr2 right before the for loop.
The implementation looks spooky if s is empty. In this case space seems to alloc an entry on the tempAllocStack containing the empty string, but where is that freed? let() ignores freeing of empty strings (?).  There seems to be a memory leak.

Reply:
I'd rather not mess with uses of `vstringdef`, because it will conflict with #38 which does a bulk translation to use the macro everywhere (and it also changes the name from `vstringdef` to `vstring_def` so added references to `vstringdef` in this PR will also cause conflicts). I think if you look you will likely find that this line is touched in #38.

Reply:
Same thing about the other suggestions: this PR is about constifying and nothing more (to the extent possible). I have a bunch more changes waiting in the wings but if I put it all in one PR it will be huge and hard to follow, so I'm trying to do one type of change at a time.

Reply:
Based on how the code generally handles these temp strings, it seems like empty strings never have to be deallocated. I assume the reason is because `let` will never create an empty string, it always sets the input to some designated `""` in static memory. (If that's not the case, then the whole `let(&x, "")` idiom is suspect, but even beyond that there are parts of the code that say that skip the `let(&x, "")` statement if `!x[0]`, which is bogus if "non-canonical" empty strings can be created in temporary storage.

Reply:
Canonicalization of empty strings is done here:

https://github.com/metamath/metamath-exe/blob/ddcd53aa5b6e9403f3692ebde8b42f0dc24aff7f/mmvstr.c#L128-L134

Reply:
In this function `nmbrCvtMToVString`, when `s` is empty, it does:
* `mstrLen = 0`
* `outputLen = -1`
* `let(&tmpStr, "")` (so `tempStr` is a pointer to `""` in static storage)
* The `if (tmpStr[0]) makeTempAlloc(tmpStr);` line is skipped
* `return tmpStr`

So it returns a canonical empty `vstring`, which is fine to be treated as a `temp_vstring` because it doesn't need to be deallocated.

Reply:
Under the assumption outputLen == -1
in mmdata.c line 1137 nmbrCvtMToVStr calls space(-1)
in mmvstr.c line 571 space(-1) calls string(-1, ' ')
in mmvstr.c line 562 string(0, ' ') calls tempAlloc(1) 
In mmvstr.c line 56: tempAlloc(1) allocates 1 byte memory
Question: when is this byte freed again?
This is important, since you strengthen the return type from vstring to temp_vstring.

Reply:
Aha, this is a bug in `string()`, it should return `""` if `n = 0`

Reply:
One should check whether nmbrSeg e. a. suffer from a similar problem.

Reply:
On further review (while preparing a separate PR for this issue), I think there is actually no bug here. The invariant is as follows:

* Temporary strings (`temp_vstring`, constructed by `tempAlloc` and freed by `freeTempAlloc`, usually via `let`) are allowed to be noncanonical empty strings, and are also allowed to have internal `\0` characters as a quick truncation operation. They should never be manually freed, because there is no easy means to remove them from the temp stack, and this is why they cannot appear in the left argument of a `let`. Static strings can also be treated as `temp_vstring`, because from client code this only means that the string should not be freed manually and one cannot assume that they will live past the next call to `freeTempAlloc`.

* Regular strings (`vstring`, constructed by `vstring_def`) can be in one of two states: empty, and pointing to the canonical empty string `""`, or nonempty and pointing to a `malloc`'d buffer with a control block at bytes -12..0 relative to the pointer. The `let(&str, value)` function can be used to change the value of a `vstring str`. `vstring` variables always start in the empty state using `vstring_def`, and they must be in the empty state when exiting a scope or else memory will leak.

As it applies to this code, we see that the invariants are upheld:

* `string` has no bug: it may sometimes create noncanonical empty strings but they are always temp-allocated.
* `nmbrCvtMToVString` uses `let(&tmpStr, space(outputLen))` correctly to copy a `temp_vstring` into a `vstring`, and calls `makeTempAlloc` to put it on the temporary stack. It will be deallocated at the next call to `let`.
* `makeTempAlloc` doesn't have a bug per se, but it has IMO the wrong precondition, and one that does not match the corresponding functions `nmbrMakeTempAlloc` and `pntrMakeTempAlloc`. It is supposed to convert a `vstring` into a `temp_vstring` and the way to do that is to put owned (nonempty) `vstring`s on the temp stack for later reclamation, and do nothing for (canonical static) empty `vstring`s since this already qualifies as a `temp_vstring`, it should not be deallocated. Currently, the nonempty check is done outside the function for `makeTempAlloc` and inside the function for `nmbrMakeTempAlloc` and `pntrMakeTempAlloc`; I will fix this in another PR.